File: app/services/singbox.py
# -*- coding: utf-8 -*-
"""Sing-Box 配置生成核心逻辑 (对齐 doc/vless.md 规格)。"""
import json
from typing import List

from app.config import ALLOWED_OVERRIDE_KEYS, TEMPLATE_PATH
from app.contracts import assert_protocol_supported, get_core, validate_node_contract
from app.models import Node, User, parse_config_override


def build_singbox_outbound(node: Node, user: User) -> dict:
    protocol = (node.protocol or "vless").lower()
    # 导出守卫：核心不支持该协议时明确报错（singbox 当前支持全部协议）
    assert_protocol_supported(get_core("singbox"), protocol)
    # 老数据兜底：tag 为空时用 node_name（新数据契约强制 tag 必填）
    tag = node.tag or node.node_name

    # 导出守卫：节点 + 用户凭证合并视图按契约全量校验
    # （uuid/password 属用户级凭证，节点 CRUD 阶段不校验，导出时统一把关）
    user_uuid = user.uuid or user.token
    user_password = user.password or user.token
    merged = {
        "tag": tag,
        "server_address": node.server_address,
        "server_port": node.server_port,
        "uuid": user_uuid,
        "password": user_password,
        "security": (node.security or "").lower(),
        "node_name": node.node_name,
        "flow": node.flow,
        "public_key": node.public_key,
        "short_id": node.short_id,
        "fingerprint": node.fingerprint,
        "congestion_control": node.congestion_control,
    }
    validate_node_contract(merged, protocol)

    outbound = {
        "type": protocol,
        "tag": tag,
        "server": node.server_address,
        "server_port": node.server_port,
    }

    if protocol == "tuic":
        outbound["uuid"] = user_uuid
        outbound["password"] = user_password
        outbound["congestion_control"] = node.congestion_control or "bbr"
        outbound["zero_rtt_handshake"] = False
        tls_config = {"enabled": True}
        if node.sni:
            tls_config["server_name"] = node.sni
        tls_config["alpn"] = ["h3"]
        outbound["tls"] = tls_config
    elif protocol == "vless":
        outbound["uuid"] = user_uuid
        outbound["flow"] = node.flow or "xtls-rprx-vision"

        if node.security in ["tls", "reality"]:
            tls_config = {"enabled": True}
            if node.sni:
                tls_config["server_name"] = node.sni

            # utls 开启
            tls_config["utls"] = {
                "enabled": True,
                "fingerprint": node.fingerprint or "chrome",
            }

            if node.security == "reality":
                tls_config["reality"] = {
                    "enabled": True,
                    "public_key": node.public_key or "",
                    "short_id": node.short_id or "",
                }
            outbound["tls"] = tls_config
    elif protocol == "anytls":
        outbound["uuid"] = user_uuid
        outbound["password"] = user_password
        if node.security in ["tls", "reality"]:
            tls_config = {"enabled": True}
            if node.sni:
                tls_config["server_name"] = node.sni
            if node.security == "reality":
                tls_config["reality"] = {
                    "enabled": True,
                    "public_key": node.public_key or "",
                    "short_id": node.short_id or "",
                }
            outbound["tls"] = tls_config

    if node.transport_type and node.transport_type != "direct":
        transport = {"type": node.transport_type}
        if node.path:
            transport["path"] = node.path
        outbound["transport"] = transport

    return outbound


def load_singbox_template() -> dict:
    """读取外置 template.json 模板文件。"""
    if TEMPLATE_PATH.exists():
        try:
            with open(TEMPLATE_PATH, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception as e:
            print(f"Error loading template.json ({e}), falling back to basic template.")

    # 基本降级模板
    return {
        "log": {"level": "info", "timestamp": True},
        "outbounds": [
            {"type": "direct", "tag": "direct", "routing_mark": 1024},
            {"tag": "Proxy", "type": "selector", "outbounds": ["urltest"]},
            {"tag": "urltest", "type": "urltest", "outbounds": []},
        ],
    }


def generate_singbox_config(nodes: List[Node], user: User) -> dict:
    active_nodes = [n for n in nodes if n.is_active]
    node_outbounds = [build_singbox_outbound(n, user) for n in active_nodes]
    node_tags = [n.node_name for n in active_nodes]

    # 读取外置 template.json
    config = load_singbox_template()
    outbounds = config.get("outbounds", [])

    # 1. 遍历模板 outbounds，将节点 tag 注入到 Proxy (selector) 与 urltest
    for ob in outbounds:
        if isinstance(ob, dict):
            tag = ob.get("tag", "")
            ob_type = ob.get("type", "")

            # selector 分组 (如 tag == "Proxy" 或 type == "selector")
            if tag == "Proxy" or ob_type == "selector":
                existing = ob.get("outbounds", [])
                for n_tag in node_tags:
                    if n_tag not in existing:
                        existing.append(n_tag)
                ob["outbounds"] = existing

            # urltest 分组 (如 tag == "urltest" 或 type == "urltest")
            elif tag == "urltest" or ob_type == "urltest":
                ob["outbounds"] = node_tags

    # 2. 将该用户的节点对象直接补充到 outbounds 数组末尾
    outbounds.extend(node_outbounds)
    config["outbounds"] = outbounds

    # 3. 应用该用户的 config_override (白名单字段整体覆盖，目前 route / dns)
    override = parse_config_override(user.config_override)
    if override:
        for key in ALLOWED_OVERRIDE_KEYS:
            if key in override:
                config[key] = override[key]

    return config

File: internal/service/singbox/singbox.go
package singbox

import (
	"encoding/json"
	"os"

	"noder/internal/config"
	"noder/internal/contract"
	"noder/internal/model"
)

func BuildSingboxOutbound(node *model.Node, user *model.User) (map[string]interface{}, error) {
	proto := node.Protocol
	if proto == "" {
		proto = "vless"
	}

	coreInfo, err := contract.GetCore("singbox")
	if err != nil {
		return nil, err
	}
	if err := contract.AssertProtocolSupported(coreInfo, proto); err != nil {
		return nil, err
	}

	tag := node.Tag
	if tag == "" {
		tag = node.NodeName
	}

	userUUID := user.Token
	if user.UUID != nil && *user.UUID != "" {
		userUUID = *user.UUID
	}
	userPassword := user.Token
	if user.Password != nil && *user.Password != "" {
		userPassword = *user.Password
	}

	// 校验全量契约
	merged := map[string]interface{}{
		"tag":            tag,
		"node_name":      node.NodeName,
		"protocol":       proto,
		"server_address": node.ServerAddress,
		"server_port":    node.ServerPort,
		"security":       node.Security,
		"uuid":           userUUID,
		"password":       userPassword,
	}
	if node.SNI != nil {
		merged["sni"] = *node.SNI
	}
	if node.PublicKey != nil {
		merged["public_key"] = *node.PublicKey
	}
	if node.ShortID != nil {
		merged["short_id"] = *node.ShortID
	}
	if node.Fingerprint != nil {
		merged["fingerprint"] = *node.Fingerprint
	}
	if node.Flow != nil {
		merged["flow"] = *node.Flow
	}
	if node.CongestionControl != nil {
		merged["congestion_control"] = *node.CongestionControl
	}

	if err := contract.ValidateNodeContract(merged, proto, false); err != nil {
		return nil, err
	}

	outbound := map[string]interface{}{
		"type":        proto,
		"tag":         tag,
		"server":      node.ServerAddress,
		"server_port": node.ServerPort,
	}

	switch proto {
	case "tuic":
		outbound["uuid"] = userUUID
		outbound["password"] = userPassword
		cc := "bbr"
		if node.CongestionControl != nil && *node.CongestionControl != "" {
			cc = *node.CongestionControl
		}
		outbound["congestion_control"] = cc
		outbound["zero_rtt_handshake"] = false
		tlsConfig := map[string]interface{}{
			"enabled": true,
			"alpn":    []string{"h3"},
		}
		if node.SNI != nil && *node.SNI != "" {
			tlsConfig["server_name"] = *node.SNI
		}
		outbound["tls"] = tlsConfig

	case "vless":
		outbound["uuid"] = userUUID
		flow := "xtls-rprx-vision"
		if node.Flow != nil && *node.Flow != "" {
			flow = *node.Flow
		}
		outbound["flow"] = flow

		if node.Security == "tls" || node.Security == "reality" {
			fp := "chrome"
			if node.Fingerprint != nil && *node.Fingerprint != "" {
				fp = *node.Fingerprint
			}
			tlsConfig := map[string]interface{}{
				"enabled": true,
				"utls": map[string]interface{}{
					"enabled":     true,
					"fingerprint": fp,
				},
			}
			if node.SNI != nil && *node.SNI != "" {
				tlsConfig["server_name"] = *node.SNI
			}
			if node.Security == "reality" {
				pk := ""
				if node.PublicKey != nil {
					pk = *node.PublicKey
				}
				sid := ""
				if node.ShortID != nil {
					sid = *node.ShortID
				}
				tlsConfig["reality"] = map[string]interface{}{
					"enabled":    true,
					"public_key": pk,
					"short_id":   sid,
				}
			}
			outbound["tls"] = tlsConfig
		}

	case "anytls":
		outbound["password"] = userPassword
		if node.Security == "tls" || node.Security == "reality" {
			tlsConfig := map[string]interface{}{
				"enabled": true,
			}
			if node.SNI != nil && *node.SNI != "" {
				tlsConfig["server_name"] = *node.SNI
			}
			if node.Security == "reality" {
				pk := ""
				if node.PublicKey != nil {
					pk = *node.PublicKey
				}
				sid := ""
				if node.ShortID != nil {
					sid = *node.ShortID
				}
				tlsConfig["reality"] = map[string]interface{}{
					"enabled":    true,
					"public_key": pk,
					"short_id":   sid,
				}
			}
			outbound["tls"] = tlsConfig
		}
	}

	if node.TransportType != "" && node.TransportType != "direct" {
		trans := map[string]interface{}{
			"type": node.TransportType,
		}
		if node.Path != nil && *node.Path != "" {
			trans["path"] = *node.Path
		}
		outbound["transport"] = trans
	}

	return outbound, nil
}

func LoadSingboxTemplate() map[string]interface{} {
	if data, err := os.ReadFile(config.SingBoxTemplatePath); err == nil && len(data) > 0 {
		var res map[string]interface{}
		if err := json.Unmarshal(data, &res); err == nil {
			return res
		}
	}
	return map[string]interface{}{
		"log": map[string]interface{}{"level": "info", "timestamp": true},
		"outbounds": []interface{}{
			map[string]interface{}{"type": "direct", "tag": "direct"},
			map[string]interface{}{"tag": "Proxy", "type": "selector", "outbounds": []interface{}{"urltest"}},
			map[string]interface{}{"tag": "urltest", "type": "urltest", "outbounds": []interface{}{}},
		},
	}
}

func GenerateSingboxConfig(nodes []*model.Node, user *model.User) (map[string]interface{}, error) {
	var activeNodes []*model.Node
	for _, n := range nodes {
		if n.IsActive {
			activeNodes = append(activeNodes, n)
		}
	}

	var nodeOutbounds []map[string]interface{}
	var nodeTags []string
	for _, n := range activeNodes {
		ob, err := BuildSingboxOutbound(n, user)
		if err != nil {
			return nil, err
		}
		nodeOutbounds = append(nodeOutbounds, ob)
		nodeTags = append(nodeTags, n.Tag)
	}

	cfg := LoadSingboxTemplate()

	var rawOutbounds []interface{}
	if obs, ok := cfg["outbounds"].([]interface{}); ok {
		rawOutbounds = obs
	}

	// 1. 注入 selector 与 urltest
	for _, raw := range rawOutbounds {
		if ob, ok := raw.(map[string]interface{}); ok {
			tag, _ := ob["tag"].(string)
			obType, _ := ob["type"].(string)

			if tag == "Proxy" || obType == "selector" {
				existingList, _ := ob["outbounds"].([]interface{})
				seen := make(map[string]bool)
				for _, e := range existingList {
					if s, ok := e.(string); ok {
						seen[s] = true
					}
				}
				for _, nt := range nodeTags {
					if !seen[nt] {
						existingList = append(existingList, nt)
						seen[nt] = true
					}
				}
				ob["outbounds"] = existingList
			} else if tag == "urltest" || obType == "urltest" {
				var tagInterfaces []interface{}
				for _, nt := range nodeTags {
					tagInterfaces = append(tagInterfaces, nt)
				}
				ob["outbounds"] = tagInterfaces
			}
		}
	}

	// 2. 节点放入 outbounds 末尾
	for _, nob := range nodeOutbounds {
		rawOutbounds = append(rawOutbounds, nob)
	}
	cfg["outbounds"] = rawOutbounds

	// 3. 应用 config_override
	if user.ConfigOverride != nil && *user.ConfigOverride != "" {
		var override map[string]interface{}
		if err := json.Unmarshal([]byte(*user.ConfigOverride), &override); err == nil {
			for k, v := range override {
				if config.AllowedOverrideKeys[k] {
					cfg[k] = v
				}
			}
		}
	}

	return cfg, nil
}
